compendium/DeclarativeServices/src/manager/BundleLoader.cpp

```diff
@@ -72,62 +75,49 @@ GetComponentCreatorDeletors(const std::string& compName,
   if (bundleBinaries.lock()->count(bundleLoc) != 0u) {
     handle = bundleBinaries.lock()->at(bundleLoc);
   } else {
-#if defined(_WIN32)
-    std::wstring bundlePathWstr = UTF8StrToWStr(fromBundle.GetLocation());
-    handle = reinterpret_cast<void*>(LoadLibraryW(bundlePathWstr.c_str()));
-    if (handle == nullptr) {
-      std::error_code err_code(GetLastError(), std::generic_category());
-      std::string errMsg("Unable to load bundle binary ");
-      errMsg += fromBundle.GetLocation();
-      errMsg += ". Error: ";
-      errMsg += std::to_string(GetLastError());
-      throw cppmicroservices::SharedLibraryException(
-        err_code, std::move(errMsg), std::move(fromBundle));
-    }
+    SharedLibrary sh(bundleLoc);
+    try {
+#ifdef US_PLATFORM_POSIX
```

Comment:
If we make a guarantee that ctx.GetProperty(Constants::LIBRARY_LOAD_OPTIONS) is always defined to its default values or possibly overwritten by customer settings, then this code collapses to this:

    try {
        auto ctx = fromBundle.GetBundleContext();
        auto opts = ctx.GetProperty(Constants::LIBRARY_LOAD_OPTIONS);
        sh.Load(opts);
    }...

Reply:
CoreBundleContext.cpp sets the libraryLoadOptions member variable in the following way:

In the constructor: 
`libraryLoadOptions(0)`

In Init():
```
#ifdef US_PLATFORM_POSIX
  try {
      libraryLoadOptions = any_cast<int>(frameworkProperties[Constants::LIBRARY_LOAD_OPTIONS]);
  } catch (...) {
      DIAG_LOG(*sink) << "Unable to read default library load options from config.";
      libraryLoadOptions = RTLD_LAZY | RTLD_LOCAL;
  }
  DIAG_LOG(*sink) << "Library Load Options = " << libraryLoadOptions;
#endif
```
It's guarding against an any_cast failure, in case the user defines a framework property value that is not an int. It does NOT guard against any integer value not defined by dlopen (i.e. -999).

Knowing this, I wanna move the try/catch block in Init() to InitProperties(...). This will parse the framework property, then if the value is an INT, it'll place the default value for POSIX. Later, Init() will assign the "correct(ed) value" to libraryLoadOptions member variable.

The question is: IF the user accidentally set framework property value to a non-integer value, is it okay for us to "correct" the value to a known working default? Or is it better to leave the framework property as the value assigned by user (although wrong), and do the checking when trying to obtain that property later (i.e. when assigning libraryLoadOptions member variable, and when trying to parse the value in BundleLoader class)?

@jeffdiclemente

Reply:
@kevinleeMW the problem (i.e. an incorrect library load option value) should be caught as early as possible and logged. In general catching and reporting the problem when it happens and not later, when its harder to reason about where the error came from, is preferable. So I suggest using a correct default value in the error case.

Reply:
POSIX:
if user defines a non-integer load option -> override with default value RTLD_LAZY | RTLD_LOCAL
if user defines an integer load option -> use the user-defined value

Windows:
load option framework property is always overriden to be 0 (whether the user defines it or not)